_source/_posts/2022-04-06-react-azure-functions.md

```diff
@@ -0,0 +1,415 @@
+
+Open `api/CreateBadge/index.js` and replace the code there with the following.
+
+```js
+const { createCanvas, loadImage } = require("canvas");
+const { rename } = require("fs");
+const querystring = require("querystring");
+
+const templateWH = [394, 225];
+const profilePictureStart = [22, 48];
+const profilePictureWH = [97, 121];
+const letterStart = [250, 205];
+
+const multipart = require("parse-multipart-data");
+
+badgeTemplateUrl = "https://i.imgur.com/HjluofW.png";
+
+const OktaJwtVerifier = require("@okta/jwt-verifier");
+
+const oktaJwtVerifier = new OktaJwtVerifier({
+  issuer: "https://{yourOktaDomain}/oauth2/default",
+});
+
+const getAuthToken = (req) => {
+  const header = req.headers["okta-authorization"];
+  const tokenParts = header.split(" ");
+  const token = tokenParts.length > 0 ? tokenParts[1] : "";
+
+  return token;
+};
+
+const drawImage = async (req) => {
+  const bodyBuffer = Buffer.from(req.body);
+  const boundary = multipart.getBoundary(req.headers["content-type"]);
+  const parts = multipart.parse(bodyBuffer, boundary);
+
+  const canvas = createCanvas(templateWH[0], templateWH[1]);
+  const ctx = canvas.getContext("2d");
+
+  // Ideally this Azure Function should call the `/userprofile` endpoint to get  
+  // the user name instead of relying on the client to send it
+  const firstLetter = parts.filter((r) => r.name === "firstLetter")[0].data.toString();
+
+  const template = await loadImage(badgeTemplateUrl);
+  ctx.drawImage(template, 0, 0, templateWH[0], templateWH[1]);
+
+  ctx.font = "68px Calibri";
+  ctx.fillStyle = "#fff";
+  ctx.fillText(firstLetter, letterStart[0], letterStart[1]);
+
+  const profileImage = await loadImage(parts[0].data);
+  ctx.drawImage(profileImage, profilePictureStart[0], profilePictureStart[1], profilePictureWH[0], profilePictureWH[1]);
+
+  return canvas;
+};
+
+module.exports = async function (context, req) {
+  const accessToken = getAuthToken(req);
+  const jwt = await oktaJwtVerifier.verifyAccessToken(accessToken, "api://default");
+
+  const canvas = await drawImage(req);
+
+  var stream = await canvas.pngStream();
+  context.res.setHeader("Content-Type", "image/png");
+  context.res.end(canvas.toBuffer("image/png"));
+};
+```
+
+This file uses the `OktaJwtVerifier` to verify the token sent from the React front end. It does this by parsing the `okta-authorization` header. If the token is invalid, it will return a 403.
+
+The other primary function of this code is to take the image uploaded by the user and modify a template image by adding the uploaded image to it. You will also pull the user's name from the JWT and replace the name on the badge with the user's first letter of their first name. Like Agent J. Assuming this was all a success, you would return the image to the SPA to display to the user.
```

Comment:
```suggestion
The other primary function of this code is to take the image uploaded by the user and modify a template image by adding the uploaded image to it. You will also pull the user's name from the JWT and replace the name on the badge with the first letter of the user's first name. Like Agent J. Assuming this was all a success, you would return the image to the SPA to display to the user.
```
clarity
Who is Agent J? Not seeing context in text or images. Does Agent J substitute for Agent N?